sdk/openai/openai-rest/samples-dev/streamChatCompletions.ts

```diff
@@ -0,0 +1,77 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+/**
+ * Demonstrates how to list chat completions for a chat context.
+ *
+ * @summary list chat completions.
+ * @azsdk-weight 100
+ */
+
+import createClient, { AzureKeyCredential, ChatCompletionsOutput } from "@azure-rest/openai";
+import { createSseStream } from "@azure/core-sse";
+
+// Load the .env file if it exists
+import * as dotenv from "dotenv";
+import { IncomingMessage } from "http";
+dotenv.config();
+
+// You will need to set these environment variables or edit the following values
+const endpoint = process.env["ENDPOINT"] || "<endpoint>";
+const azureApiKey = process.env["AZURE_API_KEY"] || "<api key>";
+
+export async function main() {
+  console.log("== Streaming Chat Completions Sample ==");
+
+  const client = createClient(endpoint, new AzureKeyCredential(azureApiKey));
+  const deploymentName = "gpt-35-turbo";
+  const response = await client
+    .path("/deployments/{deploymentId}/chat/completions", deploymentName)
+    .post({
+      body: {
+        messages: [
+          { role: "system", content: "You are a helpful assistant. You will talk like a pirate." },
+          { role: "user", content: "Can you help me?" },
+          { role: "assistant", content: "Arrrr! Of course, me hearty! What can I do for ye?" },
+          { role: "user", content: "What's the best way to train a parrot?" },
+        ],
+        stream: true,
+      },
+    })
+    .asNodeStream();
+
+  const stream = response.body;
+  if (!stream) {
+    throw new Error("The response stream is undefined");
+  }
+
+  if (response.status !== "200") {
+    throw new Error(`Failed to get chat completions: ${streamToString(stream)}`);
+  }
+
+  const sses = createSseStream(stream as IncomingMessage);
+
+  for await (const event of sses) {
+    if (event.data === "[DONE]") {
+      return;
+    }
+    for (const choice of (JSON.parse(event.data) as ChatCompletionsOutput).choices) {
+      console.log(choice.delta?.content);
+    }
+  }
+}
+
+main().catch((err) => {
+  console.error("The sample encountered an error:", err);
+});
+
+async function streamToString(stream: NodeJS.ReadableStream) {
+  // lets have a ReadableStream as a stream variable
+  const chunks = [];
+
+  for await (const chunk of stream) {
+    chunks.push(Buffer.from(chunk));
+  }
+
+  return Buffer.concat(chunks).toString("utf-8");
+}
```

Comment:
@joheredi / @xirzec / @mpodwysocki should we move this utility function to core-util so it can be easier for RLC users to access error messages when they want to stream the response?

Reply:
The problem I see is that by doing so we'd be guiding our customers to take a dependency on `@azure/code-util` in their projects which I'm not sure we want to. Having this here helps the samples be self contained.

Reply:
Yeah that is a good point. I guess the approach taken in RLCs so far is to provide convenience functions within the library so another idea is to export `streamToString` from the RLC itself where streaming scenarios are expected. Generalizing this, we may want to export a function that handles a potential error if we have a stream response:
```ts
function handleStreamError(
  response: HttpNodeStreamResponse
): asserts response is HttpNodeStreamResponse & { body: NodeJS.ReadableStream } {
  const stream = response.body;
  if (!stream) {
    throw new Error("The response stream is undefined");
  }

  if (!["200"].includes(response.status)) {
    throw new Error(`Error: ${streamToString(stream)}`);
  }
}
```

This is especially useful because `isUnexpected` is currently broken in streaming scenarios, see https://github.com/Azure/autorest.typescript/issues/2008.

There is also a question about standardizing on a web stream but we can follow the approach taken in implementing file upload streaming in form-data scenarios. /cc @timovv

Reply:
Let's get feedback on this helper by shipping it here and if we find it's generally useful for RLC customers (or at least ones that have streams) we could look at re-exporting it from a shared place (e.g. core-util) but I think we'd still want to have the consumer pattern be importing it from the client package (similar to how we re-export credentials from core-auth)